tango/integrations/transformers/soft_prompt.py

```diff
@@ -0,0 +1,248 @@
+            device=original_embedding.weight.device,
+        )
+    )
+    r = random.Random(random_seed)
+    if initialize_from_top_embeddings is None:
+        initialize_from_top_embeddings = original_embedding.num_embeddings
+    indices = torch.tensor(r.sample(range(initialize_from_top_embeddings), prompt_length))
+    with torch.no_grad():
+        prompt_embedding.copy_(original_embedding(indices).unsqueeze(0))
+
+    if only_prompt_is_trainable:
+        for param in model.parameters():
+            param.requires_grad = False
+
+    # find unique parameter name
+    parameter_name = "prompt_embedding"
+    parameter_name_index = 0
+    while True:
+        try:
+            model.get_parameter(parameter_name)
+        except AttributeError:
+            break
+        parameter_name_index += 1
+        parameter_name = f"prompt_embedding_{parameter_name_index}"
+    model.register_parameter(parameter_name, prompt_embedding)
+
+    def patch_tensor(kwargs: Dict[str, torch.Tensor], key: str, value: Any = 0) -> None:
+        t = kwargs.get(key)
+        if t is None:
+            return
+        prefix = t.new_full((t.size(0), prompt_length) + t.shape[2:], value)
+        kwargs[key] = torch.cat([prefix, t], dim=1)
+
+    def patch_tensor_with_indices(
+        kwargs: Dict[str, torch.Tensor], key: str, offset: int = 0
+    ) -> None:
+        t = kwargs.get(key)
+        if t is None:
+            return
+        kwargs[key] = torch.cat(
+            [
+                torch.arange(0, prompt_length, dtype=t.dtype)
+                .unsqueeze(0)
+                .expand(t.size(0), prompt_length),
+                t + offset,
+            ],
+            dim=1,
+        )
+
+    old_forward = model.forward
+
+    def new_forward(*args, **kwargs):
+
+        # Massage the input to include the prompt
+        if kwargs.get("past_key_values") is not None:
+            # If we have already been running this model, we don't need to do anything with the prefix now.
+            return old_forward(*args, **kwargs)
+        if kwargs.get("encoder_outputs") is not None:
+            # For encoder/decoder models, this runs only on the encoder. If we already have encoder outputs,
+            # we don't have to do anything.
+            return old_forward(*args, **kwargs)
+
+        inputs_embeds: Optional[torch.Tensor] = None
+        input_ids = kwargs.pop("input_ids", None)
+        if input_ids is not None:
+            inputs_embeds = original_embedding(input_ids)
+
+        inputs_embeds = kwargs.get("inputs_embeds", inputs_embeds)
+        if inputs_embeds is not None:
+            kwargs["inputs_embeds"] = torch.cat(
+                [prompt_embedding.expand(inputs_embeds.size(0), -1, -1), inputs_embeds], dim=1
+            )
+
+        patch_tensor(kwargs, "labels")
+        patch_tensor(kwargs, "attention_mask", 1)
+        patch_tensor(kwargs, "token_type_ids")
+        patch_tensor_with_indices(kwargs, "position_ids", prompt_length)
```

Comment:
So, if the position ids are originally `[0, 1, 2, 3, 4]`, they will now be `[0, 1, 2, .. prompt_len-1, 0, 1, 2, 3, 4]` ?

Reply:
Yes, that's right. I could see it going the other way, but I think it's important that the output does not change in the case where the soft prompt is configured to do nothing. Also, if we offset the position ids, we would decrease the max length that the model can handle, which is uncomfortable.